### PR TITLE
Configuring with pyproject.toml

### PR DESCRIPTION
**Pull request by WilliamBergamin**
These changes started out as an exploration of the `pyproject.toml` file. In the process I learnt that since [PEP 621](https://peps.python.org/pep-0621/), the Python community selected `pyproject.toml` as a standard way of specifying project metadata.

This PR aims to introduce changes that make the `pyproject.toml` the default configuration file for bolt-python, there are some slight differences but the testing, packaging and publishing behaviors should remain the same as they were

I've opted to rely on the [Requirements File Format](https://pip.pypa.io/en/latest/reference/requirements-file-format/#requirements-file-format) pattern to import test dependencies since it yields a clear distinction between the project dependencies and the development dependencies
 
Is the `pyproject.toml` configuration pattern something we would want to adopt in our python projects?

### Category

* [ ] `slack_bolt.App` and/or its core components
* [ ] `slack_bolt.async_app.AsyncApp` and/or its core components
* [ ] Adapters in `slack_bolt.adapter`
* [ ] Document pages under `/docs`
* [x] Others

## Requirements

Please read the [Contributing guidelines](https://github.com/slackapi/bolt-python/blob/main/.github/contributing.md) and [Code of Conduct](https://slackhq.github.io/code-of-conduct) before creating this issue or pull request. By submitting, you are agreeing to those rules.

* [x] I've read and understood the [Contributing Guidelines](https://github.com/slackapi/bolt-python/blob/main/.github/contributing.md) and have done my best effort to follow them.
* [x] I've read and agree to the [Code of Conduct](https://slackhq.github.io/code-of-conduct).
* [x] I've run `./scripts/install_all_and_run_tests.sh` after making the changes.

### CODE REVIEW

**Comment by seratch** [comment]
@WilliamBergamin Thanks for proposing this! As long as the release process goes well, migrating to pyproject.toml should be a good way to go! Can you adjust the CI build settings to pass the tests? Also, checking if the package upload works using https://test.pypi.org/ would be safe.

File: .github/workflows/codecov.yml [hunk 1]
```diff
@@ -28,10 +28,10 @@ jobs:
         run: |
           python setup.py install
```
Comment (seratch) [inline .github/workflows/codecov.yml:29]: Perhaps, need changes on this part

Comment (WilliamBergamin) [reply .github/workflows/codecov.yml]: Great point, I've fixed up the CI parts to work with the new changes

NOTE: there are some limitation with python 3.6 that required work arounds

**Comment by seratch** [comment]
Thank you! I just confirmed the package file uploaded to test PyPI works without any issues.

File: requirements/adapter.txt [hunk 2]
```diff
@@ -0,0 +1,22 @@
+# pip install -r requirements/adapter.txt
+# NOTE: any of async ones requires pip install -r requirements/async.txt too
+# used only under src/slack_bolt/adapter
```
Comment (seratch) [inline requirements/adapter.txt:3]:
```suggestion
# used only under slack_bolt/adapter
```

**Review by seratch — APPROVED** [review]
One nit comment but this looks great to me! After fixing the minor error in a comment, you can merge this PR.

**Comment by seratch** [comment]
Also, when you merge this, please use "Squash and merge" button to make the diffs simpler

**Pull request merged**